### PR TITLE
Bug: auction bonds and reserves auctioned are not guaranteed by the pool balance

### PR DESCRIPTION
**Pull request by grandizzy**
# Description of change
## Description of bug or vulnerability and solution
**The protocol should always guarantee that pool quote token balance `>= reserveAuction.unclaimed + auctions.totalBondEscrowed` hence the auction bonds and reserves auctioned are guaranteed.**

- transfer quote token from pool is done by `Pool._transferQuoteToken` function and can happen in following scenarios:
  - `Pool.removeQuoteToken`: cannot remove from reserves auctioned and auctions bonds. changed to cap removed amount at `reserveAuction.unclaimed + auctions.totalBondEscrowed`.
  - `Pool.withdrawBonds`: already reduce `auctions.totalBondEscrowed` by `maxAmount_` to transfer, no change. `maxAmount_ = Maths.min(maxAmount_, _getNormalizedPoolQuoteTokenBalance());` removed as balance is now guaranteed
  - `Pool.kickReserveAuction`: already reduce `reserveAuction_.unclaimed` by `kickerAward_` to transfer, no change
  - `Pool.takeReserves`: already reduce `reserveAuction_.unclaimed` by `amount_` to transfer, no change

  - `ERC20Pool.drawDebt`: cannot borrow from reserves auctioned and auctions bonds. Function was changed to revert with `InsufficientLiquidity` if amount to borrow is more than `pool balance - (reserveAuction.unclaimed + auctions.totalBondEscrowed)`. Borrower should take that limit into account when borrowing (and not consider only the pool balance)

  - `ERC721Pool.drawDebt`: cannot borrow from reserves auctioned and auctions bonds. Function was changed to revert with `InsufficientLiquidity` if amount to borrow is more than `pool balance - (reserveAuction.unclaimed + auctions.totalBondEscrowed)`. Borrower should take that limit into account when borrowing (and not consider only the pool balance)
  - `ERC721Pool.take`: the `excessQuoteToken` outflow to borrower is always less than amount paid by taker (`uint256 totalQuoteTokenAmount = result.quoteTokenAmount + result.excessQuoteToken;`), no change

- added `Pool._availableQuoteToken` to retrieve available quote amount for borrow / remove (balance minus kicked reserves and auction bonds)
- added `PoolInfoUtils.availableQuoteTokenAmount` helper to retrieve available quote amount in a certain pool
- renamed `_updateKicker` to `_updateEscrowedBonds` for clarity. Fixed `totalBondEscrowed` accumulator bug when new auction kicked (increase only by the difference needed to cover bond as kicker could already have claimable funds escrowed as bonds. Previously the entire bond size of the new kicked auction was added to global accumulator). `totalBondEscrowed` is decreased only when kicker penalized or kicker claims bond (`Pool.withdrawBond`)
- introduced `DEPOSIT_BUFFER` constant to be used as extra margin for deposits when reserves are calculated (to avoid Fenwick roundings)
- reduced optimizer to 200 runs

**Invariants**:
  - added **A7**: total bond escrowed accumulator (`AuctionsState.totalBondEscrowed`) should increase when auction is kicked with the difference needed to cover the bond and should decrease only when kicker bonds withdrawn (`Pool.withdrawBonds`). Claimable bonds should be available for withdrawal from pool at any time.  
  - added **QT3**: pool quote token balance (`Quote.balanceOf(pool)`) >= liquidation bonds (`AuctionsState.totalBondEscrowed`) + reserve auction unclaimed amount (`reserveAuction.unclaimed`)
  - updated **A2**: total bond escrowed accumulator (`AuctionsState.totalBondEscrowed`) = sum of kickers locked balances (`Kicker.locked`) + sum of kickers claimable balances (`Kicker.claimable`)
  - added panic exit scenario: pool set up with 500 lenders and 3000 borrowers, skipped 100000 days; lenders start kicking loans, borrowers repay debt

**Added failing unit and regression test**
- issue was revealed by invariant Reserves test where the reserves take transaction reverted due to insufficient balance
```
forge test -vv --match-test testReserveAuctionNoFunds
```

**Fixed baseline**:
- `testTakeCollateralAndSettleSubsetPool`: skip time so borrower accrue interest / have enough funds for lender redeem LP
- `testSettleAuctionWithoutTakes`: since remaining borrower debt after settle is 0, make sure lenders can redeem as much as they can from pool but bonds withdrawal guaranteed
- `testReserveAuctionPartiallyTaken`: make sure reserves available to take after lender redeems all their LP share

# Contract size
## Pre Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,436B  (99.43%)
  ERC20Pool                -  23,864B  (97.10%)
```
## Post Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,530B  (99.81%)
  ERC20Pool                -  23,964B  (97.51%)
```

# Gas usage
- legit increase in `drawDebt` and `removeQuoteToken` as 2 extra reads from storage done in order to load escrowed bonds and claimable reserves
## Pre Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120263          | 177421 | 162088 | 712048  | 60004   |
| bucketTake                           | 308497          | 326872 | 326872 | 345248  | 4       |
| drawDebt                             | 249810          | 294735 | 268644 | 808871  | 136003  |
| kick                                 | 147086          | 221521 | 219896 | 1031944 | 48000   |
| kickWithDeposit                      | 195881          | 273335 | 269486 | 996447  | 8000    |
| moveQuoteToken                       | 277364          | 277364 | 277364 | 277364  | 1       |
| removeQuoteToken                     | 144645          | 167188 | 171242 | 194509  | 4000    |
| repayDebt                            | 574825          | 606967 | 596863 | 755134  | 16002   |
| settle                               | 348778          | 348778 | 348778 | 348778  | 1       |
| take                                 | 80960           | 82386  | 82078  | 313490  | 15998   |
```
## Post Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120153          | 177302 | 161978 | 711807  | 60004   |
| bucketTake                           | 308418          | 326793 | 326793 | 345169  | 4       |
| drawDebt                             | 247809          | 292499 | 266643 | 802874  | 136003  |
| kick                                 | 147225          | 221624 | 220000 | 1031952 | 48000   |
| kickWithDeposit                      | 196105          | 273535 | 269686 | 996531  | 8000    |
| moveQuoteToken                       | 267621          | 267621 | 267621 | 267621  | 1       |
| removeQuoteToken                     | 143070          | 163418 | 167367 | 188729  | 4000    |
| repayDebt                            | 574533          | 606676 | 596571 | 754842  | 16002   |
| settle                               | 348675          | 348675 | 348675 | 348675  | 1       |
| take                                 | 80802           | 82228  | 81920  | 313283  | 15998   |
```

### CODE REVIEW

File: src/base/Pool.sol [hunk 1]
```diff
@@ -714,6 +714,17 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
 
+    /**
+     *  @notice Returns the quote token amount available to take loans or to be removed from pool.
+     *          Ensures claimable reserves and auction bonds are not used when taking loans.
+     */
+    function _availableQuoteToken() internal view returns (uint256 quoteAvailable_) {
+        uint256 poolBalance     = _getNormalizedPoolQuoteTokenBalance();
+        uint256 escrowedAmounts = auctions.totalBondEscrowed + reserveAuction.unclaimed;
+
+        if (poolBalance > escrowedAmounts) quoteAvailable_ = poolBalance - escrowedAmounts;
```
Comment (mattcushman) [inline src/base/Pool.sol:725]: Note: if poolBalance is less than escrowedAmount this return 0, which is correct.

File: tests/forge/invariants/base/BasicInvariants.t.sol [hunk 2]
```diff
@@ -150,11 +150,16 @@ abstract contract BasicInvariants is BaseInvariants {
         console.log("assets      -> ", assets);
         console.log("liabilities -> ", liabilities);
 
+        require(
```
Comment (mattcushman) [inline tests/forge/invariants/base/BasicInvariants.t.sol]: Why not just make this a new invariant on its own?

File: tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol [hunk 3]
```diff
@@ -205,6 +209,19 @@ contract ERC20PoolLiquidationsScaledTest is ERC20DSTestPlus {
         assertGt(auctionDebt, 0);
         assertGt(auctionCollateral, collateralDust);
         _settle();
+
+        // lender redeem their shares
+        changePrank(_lender);
+        for (uint256 i = 0; i < 4; ++i) {
+            (, uint256 bucketQuote, , , ,) = _poolUtils.bucketInfo(address(_pool), startBucketId + i);
+            if (bucketQuote != 0) _pool.removeQuoteToken(type(uint256).max, startBucketId + i);
```
Comment (mattcushman) [inline tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol]: Is the difference here with tearDown that we're not checking that all LPs were removed?

**Review by mattcushman — COMMENTED** [review]
look good generally, left a couple of questions.  Part of me still likes the idea of adding a slight multiple > 1 (like 1e18+1e9 in WAD terms) of the deposits to the liabilities.

File: tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol (hunk 3 above)
Comment (grandizzy) [reply tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol]: yes, the difference is that we don't assert lender LP zeroed after remove

File: tests/forge/invariants/base/BasicInvariants.t.sol (hunk 2 above)
Comment (grandizzy) [reply tests/forge/invariants/base/BasicInvariants.t.sol]: yeah, split in different QT invariant, will do

Comment (grandizzy) [reply tests/forge/invariants/base/BasicInvariants.t.sol]: added QT3 invariant with https://github.com/ajna-finance/contracts/pull/806/commits/93140b1c36c5c6f8a89c71ea8dc661863f8b51a9

**Comment by grandizzy** [comment]
> look good generally, left a couple of questions. Part of me still likes the idea of adding a slight multiple > 1 (like 1e18+1e9 in WAD terms) of the deposits to the liabilities.

my hesitation is that it changes baseline pretty substantially and could be hard to understand why such `1e9` (of course can be documented properly) and if I correctly understand the way it is right now does not have any side effect on the way pool functions. would like to see other's opinion too and if should change or not

**Comment by mattcushman** [comment]
> > look good generally, left a couple of questions. Part of me still likes the idea of adding a slight multiple > 1 (like 1e18+1e9 in WAD terms) of the deposits to the liabilities.
> 
> my hesitation is that it changes baseline pretty substantially and could be hard to understand why such `1e9` (of course can be documented properly) and if I correctly understand the way it is right now does not have any side effect on the way pool functions. would like to see other's opinion too and if should change or not

I'm ok with it either way.  How about we discuss in a group tmw the trade-offs to make a final decision.

**Comment by grandizzy** [comment]
> > > look good generally, left a couple of questions. Part of me still likes the idea of adding a slight multiple > 1 (like 1e18+1e9 in WAD terms) of the deposits to the liabilities.
> > 
> > 
> > my hesitation is that it changes baseline pretty substantially and could be hard to understand why such `1e9` (of course can be documented properly) and if I correctly understand the way it is right now does not have any side effect on the way pool functions. would like to see other's opinion too and if should change or not
> 
> I'm ok with it either way. How about we discuss in a group tmw the trade-offs to make a final decision.

yep, sounds good! :+1:

**Review by prat-gpt — APPROVED** [review]
LGTM!

File: src/libraries/external/BorrowerActions.sol [hunk 4]
```diff
@@ -115,13 +116,17 @@ library BorrowerActions {
         DepositsState storage deposits_,
         LoansState    storage loans_,
         PoolState calldata poolState_,
+        uint256 maxAvaialble_,
```
Comment (EdNoepel) [inline src/libraries/external/BorrowerActions.sol]: spelling: `maxAvailable_`

File: tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        address quote_,
+        address collateral_,
+        address poolInfo_,
+        address testContract_
+    ) BaseERC20PoolHandler(pool_, ajna_, quote_, collateral_, poolInfo_, 0, testContract_) {
+        numberOfBuckets = buckets.length();
+        setUp();
+    }
+
+
```
Comment (EdNoepel) [inline tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol]: nit: extra blank line

File: tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        (uint256 kickerClaimable, ) = _pool.kickerInfo(kicker); 
+
+        _actor = kicker;
+        vm.startPrank(_actor);
+        _withdrawBonds(kicker, kickerClaimable);
+        vm.stopPrank();
+    }
+
+
```
Comment (EdNoepel) [inline tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol]: nit: extra blank line

**Review by EdNoepel — APPROVED** [review]
One typographic issue; otherwise looks good.

File: src/libraries/external/BorrowerActions.sol (hunk 4 above)
Comment (grandizzy) [reply src/libraries/external/BorrowerActions.sol]: fixed in https://github.com/ajna-finance/contracts/pull/806/commits/180edb65188fa5668f7a4603a0f33d6a828b84b8

File: tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol (hunk 5 above)
Comment (grandizzy) [reply tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol]: removed in https://github.com/ajna-finance/contracts/pull/806/commits/180edb65188fa5668f7a4603a0f33d6a828b84b8

File: tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol (hunk 6 above)
Comment (grandizzy) [reply tests/forge/invariants/ERC20Pool/handlers/PanicExitERC20PoolHandler.sol]: removed in https://github.com/ajna-finance/contracts/pull/806/commits/180edb65188fa5668f7a4603a0f33d6a828b84b8

**Review by mattcushman — APPROVED** [review]
lgtm

**Pull request merged**